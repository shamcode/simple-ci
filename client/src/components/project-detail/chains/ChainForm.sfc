<template>
    {% import CodeFlask from '../../CodeFlask.sfc' %}

    <div class="project-chain-form">
        <form :onsubmit={{this$._submitForm}} :ref="formNode">
            <div class="group">
                <input name="name" type="text" required value={{name}}/>
                <span class="highlight"></span>
                <span class="bar"></span>
                <label>Name</label>
            </div>
            <div class="group">
                <CodeFlask
                    name="command"
                    value={{command}}
                />
                <span class="highlight"></span>
                <span class="bar"></span>
                <label>Command</label>
            </div>
            <div class="group">
                <button type="submit" :disabled={{dataSaving}} >
                    <span>{{saveButtonText}}</span>
                </button>
            </div>
        </form>
    </div>
</template>

<script>
    function extractValue( value ) {
        return value.trim();
    }

    function validateValue( value ) {
        return value.length > 0;
    }

    function ProjectForm( options ) {
        const onSubmit = $();

        const state = options( {
            [ $.name ]: '',
            [ $.command ]: '',
            [ $.saveButtonText ]: '',
            [ $.dataSaving ]: false,
            [ onSubmit ]() {}
        } );

        this$._submitForm = e => {
            e.preventDefault();
            const formData = new FormData( this.formNode );
            const data = {};
            let dataValid = true;
            [
                'name'
            ].forEach( ( key ) => {
                const value = extractValue( formData.get( key ) );
                dataValid = dataValid && validateValue( value );
                if ( dataValid ) {
                    data[ key ] = value;
                }
            } );
            if ( dataValid ) {
                data.command = extractValue( formData.get( 'command' ) );
                state[ onSubmit ]( data );
            }
        };
    }

    export default Component( Template, ProjectForm );
</script>
