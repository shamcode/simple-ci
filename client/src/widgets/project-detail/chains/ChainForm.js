import { options } from 'sham-ui';
import ChainFormTemplate from './ChainForm.sht';

function extractValue( value ) {
    return value.trim();
}

function validateValue( value ) {
    return value.length > 0;
}

export default class ProjectForm extends ChainFormTemplate {
    @options name = '';
    @options command = '';
    @options saveButtonText = '';
    @options dataSaving = false;
    @options onSubmit() {}

    _submitForm( e ) {
        e.preventDefault();
        const formData = new FormData( this.formNode );
        const data = {};
        let dataValid = true;
        [
            'name'
        ].forEach( ( key ) => {
            const value = extractValue( formData.get( key ) );
            dataValid = dataValid && validateValue( value );
            if ( dataValid ) {
                data[ key ] = value;
            }
        } );
        if ( dataValid ) {
            data.command = extractValue( formData.get( 'command' ) );
            this.options.onSubmit( data );
        }
    }
}
