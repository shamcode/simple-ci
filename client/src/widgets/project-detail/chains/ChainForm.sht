{% import CodeFlask from '../../CodeFlask' %}

<div class="project-chain-form">
    <form :onsubmit={{this._submitForm.bind(this)}}>
        <div class="group">
            <input name="name" type="text" required value={{name}}/>
            <span class="highlight"></span>
            <span class="bar"></span>
            <label>Name</label>
        </div>
        <div class="group">
            <CodeFlask
                name="command"
                value={{command}}
            />
            <span class="highlight"></span>
            <span class="bar"></span>
            <label>Command</label>
        </div>
        <div class="group">
            <button type="submit" :disabled={{dataSaving}} >
                <span>{{saveButtonText}}</span>
            </button>
        </div>
    </form>
</div>
