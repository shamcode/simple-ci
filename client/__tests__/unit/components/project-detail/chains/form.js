import renderer from 'sham-ui-test-helpers';
import ChainForm from '../../../../../src/components/project-detail/chains/ChainForm';
import * as directives from 'sham-ui-directives';

it( 'renders correctly', () => {
    const meta = renderer( ChainForm, {
        directives
    } );
    expect( meta.toJSON() ).toMatchSnapshot();
} );


it( 'submit form', () => {
    const onSubmitMock = jest.fn();
    const { ctx } = renderer( ChainForm, {
        saveButtonText: 'Create',
        onSubmit: onSubmitMock
    }, {
        directives
    } );

    const formData = {
        name: 'Test name',
        command: 'echo 1'
    };
    ctx.container.querySelector( '[name="name"]' ).value = formData.name;
    ctx.container.querySelector( '[name="command"]' ).value = formData.command;
    ctx.container.querySelector( '[type="submit"]' ).click();

    expect( onSubmitMock.mock.calls ).toHaveLength( 1 );
    expect( onSubmitMock.mock.calls[ 0 ] ).toHaveLength( 1 );
    const data = onSubmitMock.mock.calls[ 0 ][ 0 ];
    expect( Object.keys( data ) ).toEqual( [ 'name', 'command' ] );
    expect( data.name ).toBe( formData.name );
    expect( data.command ).toBe( formData.command );
} );

it( 'form validation (name)', () => {
    const onSubmitMock = jest.fn();
    const { ctx } = renderer( ChainForm, {
        saveButtonText: 'Create',
        onSubmit: onSubmitMock
    }, {
        directives
    } );

    ctx.container.querySelector( '[type="submit"]' ).click();
    expect( onSubmitMock.mock.calls ).toHaveLength( 0 );

    ctx.container.querySelector( '[type="submit"]' ).click();
    expect( onSubmitMock.mock.calls ).toHaveLength( 0 );
} );
