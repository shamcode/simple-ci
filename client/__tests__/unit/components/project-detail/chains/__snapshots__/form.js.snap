// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Object {
  "Options": Object {
    "ID": "component",
    "command": "",
    "dataSaving": false,
    "directives": Object {
      "EventListener": [Function],
      "UpdateOnEvent": [Function],
      "disabled": [Function],
      "onChangeUpdate": [Function],
      "onInputUpdate": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
      "ref": [Function],
    },
    "name": "",
    "onSubmit": [Function],
    "saveButtonText": "",
  },
  "html": "
<div class=\\"project-chain-form\\">
  <form>
    <div class=\\"group\\">
      <input name=\\"name\\" type=\\"text\\" required=\\"\\"><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
      <label>Name</label>
    </div>
    <div class=\\"group\\">
      <textarea wrap=\\"off\\" style=\\"display: none\\" name=\\"command\\"></textarea>
      <div class=\\"code-flask\\">
        <div class=\\"codeflask codeflask--has-line-numbers\\">
          <textarea class=\\"codeflask__textarea codeflask__flatten\\" spellcheck=\\"false\\" autocapitalize=\\"off\\" autocomplete=\\"off\\" autocorrect=\\"off\\"></textarea><pre class=\\"codeflask__pre codeflask__flatten  language-html\\"><code class=\\"codeflask__code  language-html\\"></code></pre>
          <div class=\\"codeflask__lines\\"><span class=\\"codeflask__lines__line\\">1</span></div>
        </div>
      </div>
      <!--0--><span class=\\"highlight\\"></span><span class=\\"bar\\"></span>
      <label>Command</label>
    </div>
    <div class=\\"group\\">
      <button type=\\"submit\\"><span></span></button>
    </div>
  </form>
</div>
",
}
`;
